test/reduce.cpp

```diff
@@ -312,6 +318,969 @@ CPP_REDUCE_TESTS(anyTrue, any_true, float, unsigned char);
+                                   tests[t * 2 + 1].end());
+
+        // Run sum
+        ASSERT_EQ(AF_SUCCESS,
+                  af_sum_by_key(&outKeys, &outVals, inKeys, inVals, 0));
+
+        dim_t ok0, ok1, ok2, ok3;
+        dim_t ov0, ov1, ov2, ov3;
+        af_get_dims(&ok0, &ok1, &ok2, &ok3, outKeys);
+        af_get_dims(&ov0, &ov1, &ov2, &ov3, outVals);
+
+        // Get result
+        vector<int> outKeysVec(ok0 * ok1 * ok2 * ok3);
+        vector<float> outValsVec(ov0 * ov1 * ov2 * ov3);
+
+        ASSERT_EQ(AF_SUCCESS,
+                  af_get_data_ptr((void *)&outKeysVec.front(), outKeys));
+        ASSERT_EQ(AF_SUCCESS,
+                  af_get_data_ptr((void *)&outValsVec.front(), outVals));
+
+        size_t nElems = currGoldKeys.size();
+        if (std::equal(currGoldKeys.begin(), currGoldKeys.end(),
+                       outKeysVec.begin()) == false) {
+            for (size_t elIter = 0; elIter < nElems; ++elIter) {
+                EXPECT_NEAR(currGoldKeys[elIter], outKeysVec[elIter], 1e-4)
+                    << "at: " << elIter << endl;
+                EXPECT_NEAR(currGoldVals[elIter], outValsVec[elIter], 1e-4)
+                    << "at: " << elIter << endl;
+            }
+            for (int i = 0; i < (int)nElems; i++) {
+                cout << currGoldKeys[i] << ":" << currGoldVals[i] << ", ";
+            }
+
+            for (int i = 0; i < (int)nElems; i++) {
+                cout << outKeysVec[i] << ":" << outValsVec[i] << ", ";
+            }
+            FAIL();
+        }
+
+        ASSERT_EQ(AF_SUCCESS, af_release_array(outKeys));
+        ASSERT_EQ(AF_SUCCESS, af_release_array(outVals));
+        ASSERT_EQ(AF_SUCCESS, af_release_array(inKeys));
+        ASSERT_EQ(AF_SUCCESS, af_release_array(inVals));
+    }
+}
+TEST(ReduceByKey, MultiBlockReduceContig10) {
+    reduce_by_key_test(string(TEST_DIR "/reduce/test_contig10_by_key.test"));
+}
+
+TEST(ReduceByKey, MultiBlockReduceRandom10) {
+    reduce_by_key_test(string(TEST_DIR "/reduce/test_random10_by_key.test"));
+}
+
+TEST(ReduceByKey, MultiBlockReduceContig500) {
+    reduce_by_key_test(string(TEST_DIR "/reduce/test_contig500_by_key.test"));
+}
+
+TEST(ReduceByKey, MultiBlockReduceByKeyRandom500) {
+    reduce_by_key_test(string(TEST_DIR "/reduce/test_random500_by_key.test"));
+}
+
+TEST(ReduceByKey, productReduceByKey) {
+    const static int testSz      = 8;
+    const int testKeys[testSz]   = {0, 2, 2, 9, 5, 5, 5, 8};
+    const float testVals[testSz] = {0, 7, 1, 6, 2, 5, 3, 4};
+
+    array keys(testSz, testKeys);
+    array vals(testSz, testVals);
+
+    array reduced_keys, reduced_vals;
+    productByKey(reduced_keys, reduced_vals, keys, vals, 0, 1);
+
+    const int goldSz                = 5;
+    const float gold_reduce[goldSz] = {0, 7, 6, 30, 4};
+    float *h_a                      = reduced_vals.host<float>();
+
+    for (int i = 0; i < goldSz; i++) { ASSERT_EQ(gold_reduce[i], h_a[i]); }
```

Comment:
Maybe not this time around, but a todo in any case: have these kinds of tests take advantage of `ASSERT_ARRAYS_EQ` or `ASSERT_VEC_ARRAY_EQ`, to better examine any potential test failures in the future. I see that other tests do take advantage of it, but maybe these ones were created before the new assert functions were merged to master.